src/commons/build-partial-update-state-command.ts

```diff
@@ -65,18 +77,53 @@ export const buildPartialUpdateAccountStateCommand = (
   baseUpdateItemCommandInput['ExpressionAttributeValues'][':sa'] = { N: `${eventTimestamp}` };
   const stringBuilder = new HistoryStringBuilder();
   baseUpdateItemCommandInput['ExpressionAttributeNames']['#H'] = 'history';
-  baseUpdateItemCommandInput['ExpressionAttributeValues'][':empty_list'] = { L: [] };
   baseUpdateItemCommandInput['ExpressionAttributeValues'][':h'] = {
-    L: [{ S: stringBuilder.getHistoryString(interventionEvent, eventTimestamp) }],
+    L: [
+      ...extractValidHistoryItems(historyList),
+      { S: stringBuilder.getHistoryString(interventionEvent, eventTimestamp) },
+    ],
   };
-  baseUpdateItemCommandInput['UpdateExpression'] +=
-    ', #INT = :int, #SA = :sa, #AA = :aa, #H = list_append(if_not_exists(#H, :empty_list), :h)';
-  if (finalState.resetPassword && finalState.reproveIdentity) {
-    baseUpdateItemCommandInput['UpdateExpression'] += ' REMOVE resetPasswordAt, reprovedIdentityAt';
-  } else if (finalState.resetPassword && !finalState.reproveIdentity) {
-    baseUpdateItemCommandInput['UpdateExpression'] += ' REMOVE resetPasswordAt';
-  } else if (!finalState.resetPassword && finalState.reproveIdentity) {
-    baseUpdateItemCommandInput['UpdateExpression'] += ' REMOVE reprovedIdentityAt';
-  }
+  baseUpdateItemCommandInput['UpdateExpression'] += ', #INT = :int, #SA = :sa, #AA = :aa, #H = :h';
+  baseUpdateItemCommandInput['UpdateExpression'] += buildRemoveExpression(finalState);
+
   return baseUpdateItemCommandInput;
 };
+
+/**
+ * Helper function to build the Remove Expression for DynamoDB update
+ * @param finalState - new account state object
+ */
+function buildRemoveExpression(finalState: StateDetails) {
+  const itemsToRemove = [];
+
+  if (finalState.resetPassword) {
+    itemsToRemove.push('resetPasswordAt');
+  }
+  if (finalState.reproveIdentity) {
+    itemsToRemove.push('reprovedIdentityAt');
+  }
+
+  if (itemsToRemove.length === 0) {
+    return '';
+  }
+
+  return ' REMOVE ' + itemsToRemove.join(', ');
+}
+
+/**
+ * Helper function to determine which history items have not exceeded the retention period.
+ * @param historyList - list of history items
+ */
+function extractValidHistoryItems(historyList: string[]) {
+  const historyStringBuilder = new HistoryStringBuilder();
+  const retentionPeriod = AppConfigService.getInstance().historyRetentionSeconds;
+  const currentTimeSeconds = getCurrentTimestamp().seconds;
+
+  return historyList.map((historyItem) => {
```

Comment:
I may be wrong, but I this that map here will return an array of `{ S: string }` mixed with undefined (when the condition is not met), I think those are ignored downstream, but may be worth using a `reduce` instead so we filter + map at the same time, to avoid any unexpected behaviour?